src/napi.h

```diff
@@ -0,0 +1,663 @@
+    Persistent(Persistent&& other);
+    Persistent& operator =(Persistent&& other);
+    Persistent& operator =(Persistent&) = delete;
+    Persistent(const Persistent&) = delete;
+
+    operator napi_persistent() const;
+
+    Env Env() const;
+    Value Value() const;
+
+  private:
+    napi_env _env;
+    napi_persistent _persistent;
+  };
+
+  class CallbackInfo {
+  public:
+    CallbackInfo(napi_env env, napi_callback_info info);
+
+    Env Env() const;
+    int Length() const;
+    const Value operator [](int index) const;
+    Object This() const;
+    void* Data() const;
+
+  private:
+    napi_env _env;
+    napi_value _this;
+    std::vector<napi_value> _args;
+    void* _data;
+  };
+
+  class PropertyDescriptor {
+  public:
+    PropertyDescriptor(napi_property_descriptor desc) : _desc(desc) {}
+
+    operator napi_property_descriptor&() { return _desc; }
+    operator const napi_property_descriptor&() const { return _desc; }
+
+  private:
+    napi_property_descriptor _desc;
+  };
+
+  /*
+  * Property descriptor for use with ObjectWrap<T>::DefineClass(). This is different from
+  * the standalone PropertyDescriptor because it is specific to each ObjectWrap<T>
+  * subclass. This prevents using descriptors from a different class when defining a new class
+  * (preventing the callbacks from having incorrect `this` pointers).
+  */
+  template <typename T>
+  class ClassPropertyDescriptor {
+  public:
+    ClassPropertyDescriptor(napi_property_descriptor desc) : _desc(desc) {}
+
+    operator napi_property_descriptor&() { return _desc; }
+    operator const napi_property_descriptor&() const { return _desc; }
+
+  private:
+    napi_property_descriptor _desc;
+  };
+
+  /*
+   * Base class to be extended by C++ classes exposed to JavaScript.
+   * Each C++ class instance gets "wrapped" by a JavaScript object.
+   *
+   * At initialization time, the DefineClass() method must be used to
+   * hook up the accessor and method callbacks. It takes a list of
+   * property descriptors, which can be constructed via the various
+   * static methods on the base class.
+   *
+   * Example:
+   *
+   *   class Example: public Napi::ObjectWrap<Example> {
+   *   public:
+   *     static void Initialize(Napi::Env& env, Napi::Object& target) {
+   *       Napi::Function constructor = DefineClass(env, "Example", New, {
+   *         InstanceAccessor("value", &GetValue, &SetValue),
+   *         InstanceMethod("doSomething", &DoSomething),
+   *       });
+   *       target.Set("Example", constructor);
+   *     }
+   *
+   *     static Example* New(const Napi::CallbackInfo& info) {
+   *       return new Example();
+   *     }
+   *
+   *     Napi::Value GetValue(const Napi::CallbackInfo& info);
+   *     void SetValue(const Napi::CallbackInfo& info, const Napi::Value& value);
+   *     Napi::Value DoSomething(const Napi::CallbackInfo& info);
+   *   }
+   */
+  template <typename T>
+  class ObjectWrap {
```

Comment:
This class needs `Ref()` and `Unref()` methods.

I have implemented those methods locally using a combination of NAPI weak references and persistent references, but it could be done better using some [NAPI referencing improvements I proposed](https://github.com/nodejs/abi-stable-node/issues/93). And then with those changes, I would probably change the `Persistent` class above to `Reference<T>`, and have `ObjectWrap<T>` extend `Reference<Object>` so it would inherit the `Ref()` / `Unref()` methods from it.